File: extensions/aether-vault/kotlin/src/main/kotlin/vault/VaultHealth.kt
package vault

import java.time.Instant
import java.util.UUID

data class VaultHealth(
    val manifestId: UUID,
    val totalShards: Int,
    val availableShards: Int,
    val parityShards: Int,
    val availableParityShards: Int,
    val minShardsForRecovery: Int,
    val replicationFactor: Int,
    val degradedNodes: List<String> = emptyList(),
    val lastCheckedAt: Instant = Instant.now()
) {
    val isRecoverable: Boolean
        get() = availableShards >= minShardsForRecovery

    val isHealthy: Boolean
        get() = availableShards == totalShards && availableParityShards == parityShards

    val isDegraded: Boolean
        get() = isRecoverable && !isHealthy

    val healthPercent: Double
        get() = if (totalShards == 0) 0.0
                else (availableShards.toDouble() / totalShards.toDouble()) * 100.0
}

File: extensions/aether-vault/swift/Sources/AetherVault/VaultHealth.swift
import Foundation

public struct VaultHealth: Codable, Equatable {
    public let manifestId: UUID
    public let totalShards: Int
    public let availableShards: Int
    public let parityShards: Int
    public let availableParityShards: Int
    public let minShardsForRecovery: Int
    public let replicationFactor: Int
    public var degradedNodes: [String]
    public let lastCheckedAt: Date

    public init(
        manifestId: UUID,
        totalShards: Int,
        availableShards: Int,
        parityShards: Int,
        availableParityShards: Int,
        minShardsForRecovery: Int,
        replicationFactor: Int,
        degradedNodes: [String] = [],
        lastCheckedAt: Date = Date()
    ) {
        self.manifestId = manifestId
        self.totalShards = totalShards
        self.availableShards = availableShards
        self.parityShards = parityShards
        self.availableParityShards = availableParityShards
        self.minShardsForRecovery = minShardsForRecovery
        self.replicationFactor = replicationFactor
        self.degradedNodes = degradedNodes
        self.lastCheckedAt = lastCheckedAt
    }

    public var isRecoverable: Bool {
        availableShards >= minShardsForRecovery
    }

    public var isHealthy: Bool {
        availableShards == totalShards && availableParityShards == parityShards
    }

    public var isDegraded: Bool {
        isRecoverable && !isHealthy
    }

    public var healthPercent: Double {
        guard totalShards > 0 else { return 0.0 }
        return (Double(availableShards) / Double(totalShards)) * 100.0
    }
}
